### ISSUE TITLE
Reduce number of (re)allocations

### ISSUE THREAD

**Issue by Nashtare** [0xPolygonZero/proof-protocol-decoder#19]
There are several allocations / reallocations across the decoder that could probably be alleviated with some refactoring.

To be tackled after #6 probably...

**Comment by dvdplm** [comment]
@Nashtare Can you elaborate a bit on what kind of allocation strategy you would like to see here, or point to examples in the code where you'd like to see refactorings? As it stands the description is a bit vague.

**Comment by Nashtare** [comment]
There are a few TODOs / collect that could be modified, though I mostly put it as a tracking ticket, following an offline discussion with @BGluth

**Comment by BGluth** [comment]
Yeah just to add, there are a handful of unnecessary heap allocations (IIRC mostly to deal with some lifetime issues), but I would expect only extremely negligible performance improvement after dealing with all of them. It might be more to just make the codebase cleaner, since we spend almost all of our time inside `plonky2` generating proofs.

**Comment by Nashtare** [comment]
Yeah I also think it's nice to reduce these AMAP, even though they aren't performance critical compared to the prover. I also have a WIP branch for this purpose on the plonky2 side.

**Comment by dvdplm** [comment]
AMAP? You mean ASAP?

**Comment by Nashtare** [comment]
As much as possible :) But it could also be ASAAMAP combining both :D

**Comment by dvdplm** [comment]
TIL!
